### PR TITLE
14729 Move background tasks list from admin UI to Primary UI

### PR DESCRIPTION
**Pull request by arthanson**
### Fixes: #14729 

Adds list and drill-down to detail view for RQ queues:

Monosnap Background Tasks | NetBox 2024-01-17 07-52-17
Monosnap Job | NetBox 2024-01-17 09-24-53
Monosnap Queued Jobs in default | NetBox 2024-01-17 07-52-31

Added requeue, stop, enqueue and delete functionality to the detail view.

### CODE REVIEW

File: netbox/core/tables/tasks.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+from netbox.tables import NetBoxTable, columns
+from utilities.templatetags.helpers import annotated_date
+from ..models import Job
+
+
+class BackgroundTasksTable(tables.Table):
+    name = tables.LinkColumn("core:background_tasks_queues", args=[A("index")], verbose_name=_("Name"))
+    jobs = tables.LinkColumn("core:background_tasks_queues", args=[A("index")], verbose_name=_("Queued Jobs"))
+    oldest_job_timestamp = tables.Column(verbose_name=_("Oldest Queued Job"))
+    started_jobs = tables.Column(verbose_name=_("Active Jobs"))
```
Comment (jeremystretch) [inline netbox/core/tables/tasks.py:16]: We need to be more careful about using "tasks" versus "jobs." This view is showing RQ tasks, and the column headers should reflect that. (IMO it's reasonable to omit the second word and just show e.g. "Active" or "Deferred.")

File: netbox/core/tables/tasks.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    ended_at = tables.Column(verbose_name=_("Ended"))
+    status = tables.Column(empty_values=(), verbose_name=_("Status"))
+    callable = tables.Column(empty_values=(), verbose_name=_("Callable"))
+
+    class Meta:
+        attrs = {
+            'class': 'table table-hover object-list',
+        }
+
+    def render_id(self, value, record):
```
Comment (jeremystretch) [inline netbox/core/tables/tasks.py:47]: Hyperlinks should be handled by passing `linkify` to the column. This ensures that the raw data remains available for clean export.

File: netbox/core/urls.py [hunk 3]
```diff
@@ -25,6 +25,11 @@
     path('jobs/<int:pk>/', views.JobView.as_view(), name='job'),
     path('jobs/<int:pk>/delete/', views.JobDeleteView.as_view(), name='job_delete'),
 
+    # Background Tasks
+    path('background-tasks/', views.BackgroundTasksListView.as_view(), name='background_tasks_list'),
+    path('background-tasks/queues/<int:queue_index>/', views.BackgroundTasksQueueListView.as_view(), name='background_tasks_queues'),
+    path('background-tasks/queues/<int:queue_index>/<str:job_id>/', views.BackgroundTasksJobDetailView.as_view(), name='background_tasks_job_detail'),
```
Comment (jeremystretch) [inline netbox/core/urls.py:31]: We can simplifying the URL tree a bit:

* `background-queues/` - Queue list
* `background-queues/<id>/` - Tasks list for a specific queue
* `background-tasks/<id>/` - Task detail view

**Review by jeremystretch — CHANGES_REQUESTED** [review]
Thanks @arthanson! I haven't really looked at the templates yet as I know they'll need updating for #12128 anyway.

**Review by kkthxbye-code — COMMENTED** [review]
Seems to be missing both the delete and requeue actions. Especially the requeue action is useful for webhooks failing because of intermittent issues. The delete action can be useful in rare cases where a very large amout of webhooks were unnecessarily enqueued because of a bulk action or in rare cases where the Job object of a scheduled Job is deleted, but the backing task was not.

#14729 doesn't mention not porting that part of the UI. If not an oversight, this should be documented in both #14729 and the release notes.

File: netbox/core/tables/tasks.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    class Meta:
+        attrs = {
+            'class': 'table table-hover object-list',
+        }
+
+
+class BackgroundTasksQueueTable(tables.Table):
+    # id = tables.LinkColumn("core:background_tasks_queues", args=[A("index")], verbose_name=_("ID"))
+    id = tables.Column(empty_values=(), verbose_name=_("ID"))
```
[the remark this answers is not in the document]
Comment (arthanson) [reply netbox/core/tables/tasks.py:35]: yes, otherwise render doesn't get called

File: netbox/core/views.py [hunk 5]
```diff
@@ -232,3 +237,61 @@ def post(self, request, pk):
...
+# Background Tasks (RQ)
+#
+
+
+class BackgroundQueuesListView(LoginRequiredMixin, View):
+
+    def get(self, request):
+        table = tables.BackgroundQueueTable(get_statistics(run_maintenance_tasks=True)["queues"])
+        table.configure(request)
+        return render(request, 'core/background_tasks.html', {
```
Comment (jeremystretch) [inline netbox/core/views.py:251]: The template name should match the view name, e.g. `queue_list.html`.

File: netbox/core/views.py [hunk 6]
```diff
@@ -232,3 +237,61 @@ def post(self, request, pk):
         messages.success(request, f"Restored configuration revision #{pk}")
 
         return redirect(candidate_config.get_absolute_url())
+
+#
+# Background Tasks (RQ)
+#
+
+
+class BackgroundQueuesListView(LoginRequiredMixin, View):
```
Comment (jeremystretch) [inline netbox/core/views.py:246]: List views should use the singular form of the object name for consistency.
```suggestion
class BackgroundQueueListView(LoginRequiredMixin, View):
```

File: netbox/core/views.py [hunk 7]
```diff
@@ -232,3 +237,61 @@ def post(self, request, pk):
...
+
+    def get(self, request):
+        table = tables.BackgroundQueueTable(get_statistics(run_maintenance_tasks=True)["queues"])
+        table.configure(request)
+        return render(request, 'core/background_tasks.html', {
+            'table': table,
+        })
+
+
+class BackgroundTasksListView(LoginRequiredMixin, View):
```
Comment (jeremystretch) [inline netbox/core/views.py:256]:
```suggestion
class BackgroundTaskListView(LoginRequiredMixin, View):
```

File: netbox/core/views.py [hunk 8]
```diff
@@ -232,3 +237,61 @@ def post(self, request, pk):
...
+
+        table = tables.BackgroundTaskTable(data=jobs, queue_index=queue_index)
+        table.configure(request)
+        return render(request, 'core/background_tasks_queue.html', {
+            'table': table,
+            'queue': queue,
+        })
+
+
+class BackgroundTaskDetailView(LoginRequiredMixin, View):
```
Comment (jeremystretch) [inline netbox/core/views.py:275]:
```suggestion
class BackgroundTaskView(LoginRequiredMixin, View):
```

File: netbox/core/tables/tasks.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        )
+
+    def render_created_at(self, value, record):
+        return annotated_date(value)
+
+    def render_enqueued_at(self, value, record):
+        return annotated_date(value)
+
+    def render_ended_at(self, value, record):
+        return annotated_date(value)
```
Comment (jeremystretch) [inline netbox/core/tables/tasks.py:73]: Are these necessary? Elsewhere in the application we use `DateTimeColumn` without any custom rendering.

File: netbox/core/tables/tasks.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        empty_text = _('No queues found')
+        fields = (
+            'id', 'created_at', 'enqueued_at', 'ended_at', 'status', 'callable',
+        )
+        default_columns = (
+            'id', 'created_at', 'enqueued_at', 'ended_at', 'status', 'callable',
+        )
+        attrs = {
+            'class': 'table table-hover object-list',
+        }
```
Comment (jeremystretch) [inline netbox/core/tables/tasks.py:58]: This is set on BaseTable.
```suggestion
```

File: netbox/core/tables/tasks.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        empty_text = _('No tasks found')
+        fields = (
+            'name', 'jobs', 'oldest_job_timestamp', 'started_jobs', 'deferred_jobs', 'finished_jobs', 'failed_jobs', 'scheduled_jobs', 'workers', 'host', 'port', 'db', 'pid',
+        )
+        default_columns = (
+            'name', 'jobs', 'started_jobs', 'deferred_jobs', 'finished_jobs', 'failed_jobs', 'scheduled_jobs',
+        )
+        attrs = {
+            'class': 'table table-hover object-list',
+        }
```
Comment (jeremystretch) [inline netbox/core/tables/tasks.py:37]:
```suggestion
```

File: netbox/core/tables/tasks.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
+import django_tables2 as tables
+from django_tables2.utils import A  # alias for Accessor
+from django.urls import reverse
+from django.utils.html import mark_safe
+from django.utils.translation import gettext_lazy as _
+
+from netbox.tables import BaseTable, NetBoxTable, columns
+from utilities.templatetags.helpers import annotated_date
+from ..models import Job
```
Comment (jeremystretch) [inline netbox/core/tables/tasks.py:9]: Remove unused imports.
```suggestion
from netbox.tables import BaseTable
from utilities.templatetags.helpers import annotated_date
```

File: netbox/core/tables/tasks.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+            'class': 'table table-hover object-list',
+        }
+
+
+class BackgroundTaskTable(BaseTable):
+    id = tables.Column(empty_values=(), verbose_name=_("ID"))
+    created_at = tables.DateTimeColumn(verbose_name=_("Created"))
+    enqueued_at = tables.DateTimeColumn(verbose_name=_("Enqueued"))
+    ended_at = tables.DateTimeColumn(verbose_name=_("Ended"))
+    status = tables.Column(empty_values=(), verbose_name=_("Status"))
```
Comment (jeremystretch) [inline netbox/core/tables/tasks.py:45]: This should render a colored badge similar to e.g. `status` on the Site model.

File: netbox/core/tables/tasks.py [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        return annotated_date(value)
+
+    def render_enqueued_at(self, value, record):
+        return annotated_date(value)
+
+    def render_ended_at(self, value, record):
+        return annotated_date(value)
+
+    def render_status(self, value, record):
+        return record.get_status
```
Comment (jeremystretch) [inline netbox/core/tables/tasks.py:76]: I think we can avoid defining a custom render method by specifying an accessor on the column.

File: netbox/core/views.py [hunk 15]
```diff
@@ -232,3 +237,61 @@ def post(self, request, pk):
...
+
+    def get(self, request, queue_index, job_id):
+        queue_index = int(queue_index)
+        queue = get_queue_by_index(queue_index)
+
+        try:
+            job = RQ_Job.fetch(job_id, connection=queue.connection, serializer=queue.serializer)
+        except NoSuchJobError:
+            raise Http404("Couldn't find job with this ID: %s" % job_id)
```
Comment (jeremystretch) [inline netbox/core/views.py:284]: Error message needs to be wrapped with `gettext()`. Could also probably make it shorter, e.g. "Job {ID} not found."

File: netbox/core/views.py [hunk 16]
```diff
@@ -232,3 +237,61 @@ def post(self, request, pk):
...
+
+        try:
+            job.func_name
+            data_is_valid = True
+        except Exception:
+            data_is_valid = False
```
Comment (jeremystretch) [inline netbox/core/views.py:290]: Not sure what's going on here. Needs an explanatory comment.

File: netbox/core/views.py [hunk 17]
```diff
@@ -232,3 +237,61 @@ def post(self, request, pk):
...
+        table.configure(request)
+        return render(request, 'core/background_tasks.html', {
+            'table': table,
+        })
+
+
+class BackgroundTasksListView(LoginRequiredMixin, View):
+
+    def get(self, request, queue_index):
+        queue_index = int(queue_index)
```
Comment (jeremystretch) [inline netbox/core/views.py:259]: `queue_index` is already passed as an integer by the URL path definition.
```suggestion
```

**Review by jeremystretch — CHANGES_REQUESTED** [review]

**Comment by jeremystretch** [comment]
@arthanson could you please add tests for the new views? Happy to do a final pass once those are in place.

**Comment by arthanson** [comment]
@jeremystretch tests added.

**Review by jeremystretch — APPROVED** [review]
@arthanson I took the liberty of doing some cleanup. There's a lot of opportunity to add more functionality, but IMO this meets the goal of #14729 and I'd rather not hold this up any longer. Please have a look at my changes from today; if you have no objections this is ready to be merged. Thanks!

**Comment by arthanson** [comment]
@jeremystretch looks good to merge

**Pull request merged**